controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -300,49 +332,53 @@ func (r *KubeadmControlPlaneReconciler) initializeControlPlane(ctx context.Conte
 		ownerRef,
 	)
 	if err != nil {
-		return errors.Wrap(err, "Failed to generate bootstrap config")
+		errs = append(errs, errors.Wrap(err, "Failed to generate bootstrap config"))
 	}
 
-	// Create the Machine
-	err = r.MachineGenerator.GenerateMachine(
-		ctx,
-		r.Client,
-		kcp.Namespace,
-		kcp.Name,
-		cluster.Name,
-		kcp.Spec.Version,
-		infraRef,
-		bootstrapRef,
-		generateKubeadmControlPlaneLabels(cluster.Name),
-		ownerRef,
-	)
-	if err != nil {
-		logger.Error(err, "Unable to create Machine")
-		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "FailedCreateMachine", "Failed to create machine: %v", err)
-
-		errs := []error{err}
+	// Only proceed to generating the Machine if we haven't encountered an error
+	if len(errs) == 0 {
+		err := r.MachineGenerator.GenerateMachine(
+			ctx,
+			r.Client,
+			kcp.Namespace,
+			kcp.Name,
+			cluster.Name,
+			kcp.Spec.Version,
+			infraRef,
+			bootstrapRef,
+			generateKubeadmControlPlaneLabels(cluster.Name),
+			ownerRef,
+		)
```

Comment:
This is rather a large amount of input parameters, I'd suggest to use something like `GenerateMachineInput` as struct to pass in these fields.

The client seems could be passed in during initialization of the `MachineGenerator`?

Reply:
Sure, I can work up a separate PR to clean this up a bit.

Reply:
I can help here as well if that's ok with you!

Reply:
Already working something up :)

Reply:
Cleaned up the argument list.